### ISSUE TITLE
Completion waiting dots cause prompt to disappear with multi line prompts

### ISSUE THREAD

**Issue by chrishoage** [ohmyzsh/ohmyzsh#5765]
With completion waiting dots off:
![screen shot 2017-01-09 at 1 47 51 pm](https://cloud.githubusercontent.com/assets/1056886/21784998/b15156e8-d672-11e6-8699-b8fd80de909f.png)

With completion waiting dots on:
![screen shot 2017-01-09 at 1 48 08 pm](https://cloud.githubusercontent.com/assets/1056886/21784997/b1508fb0-d672-11e6-9ebe-67313db39311.png)

This issue is not specific to my custom theme. I have tried other multi line prompts and all have the same issue. 

In the meantime I have disabled `COMPLETION_WAITING_DOTS`

**Comment by daneah** [comment]
I didn't experience this issue until upgrading to macOS High Sierra. I asked [this StackOverflow question](https://stackoverflow.com/questions/46440121/different-multiline-prompt-behavior-using-zsh-oh-my-zsh-iterm2-high-sierra) before finally tracing the issue to `COMPLETION_WAITING_DOTS` myself.

**Comment by mcornella** [comment]
Can you test with this modification?
```
expand-or-complete-with-dots() {
    print -Pn "%{%F{red}......%f%}"

    zle expand-or-complete
    zle redisplay
}
zle -N expand-or-complete-with-dots
bindkey '^I' expand-or-complete-with-dots
```

**Comment by daneah** [comment]
@mcornella I've just tried this and I'm still seeing the same behavior. I believe the `zle redisplay` is what causes this affliction, as commenting it out makes things behave roughly as I'd expect.

**Comment by mcornella** [comment]
Sadly I can't reproduce this. Can you share your theme so I can try on my end? Also, what `zsh --version` are you running?

**Comment by daneah** [comment]
I'm on `zsh 5.5.1 (x86_64-apple-darwin17.5.0)` on macOS 10.13.4.

My theme is a fairly heavily customized one, originally started from `kolo`. Here's [the gist](https://gist.github.com/daneah/6b85b4aabb4bbc6e806b0c443bfcf350).

**Comment by daneah** [comment]
I attempted to make an `asciinema` recording of this behavior to help you see a little more clearly what's happening, but `asciinema` drops you into a subshell where it seems that the completion dots don't execute (and therefore the behavior is not exhibited). I don't quite understand why that is, but sorry if that observation is just noise 😅

**Comment by shilov** [comment]
@mcornella just tried your modification and the issue is still present.

Also running zsh 5.5.1, but with the [denysdovhan/spaceship-prompt](https://github.com/denysdovhan/spaceship-prompt) theme.
As @daneah pointed out, the issue seems to be related to `zle redisplay`

**Comment by mcornella** [comment]
I can't reproduce the issue with either of your themes even with the same zsh version.

These are two modified versions, please try both and report if the issue persists (you can copy-paste directly into the terminal):

#### Version 1
```
expand-or-complete-with-dots() {
    # toggle line-wrapping off and back on again
    [[ -n "$terminfo[rmam]" && -n "$terminfo[smam]" ]] && echoti rmam
    print -Pn "%{%F{red}......%f%}"
    [[ -n "$terminfo[rmam]" && -n "$terminfo[smam]" ]] && echoti smam

    zle expand-or-complete
    zle -R
}
zle -N expand-or-complete-with-dots
bindkey "^I" expand-or-complete-with-dots
```

#### Version 2
```
expand-or-complete-with-dots() {
    # toggle line-wrapping off and back on again
    [[ -n "$terminfo[rmam]" && -n "$terminfo[smam]" ]] && echoti rmam
    print -Pn "%{%F{red}......%f%}"
    [[ -n "$terminfo[rmam]" && -n "$terminfo[smam]" ]] && echoti smam

    zle expand-or-complete
    zle reset-prompt
}
zle -N expand-or-complete-with-dots
bindkey "^I" expand-or-complete-with-dots
```

**Comment by daneah** [comment]
@mcornella I tried these both.

The first approach no longer overwrites the terminal prompt, but does result in the dots interrupting the line currently being typed:

`ls /va.....ar/li.....ib/...../postfix/`

The second approach works well unless I hit `<TAB> <TAB>` to cycle through argument options, i.e.:

```
ls /va<TAB>
ls /va...
ls /var/
ls /var/<TAB>
```

**Comment by daneah** [comment]
Since you can't reproduce this I'm wondering if a particular plugin is to blame. I will try keeping completion dots on and turning off various plugins to see if I can narrow that down.

**Comment by juanibiapina** [comment]
I have this problem as well. I don't use oh my zsh, but copied over just the `COMPLETION_WAITING_DOTS` part. My prompt is also custom with two lines (and right prompt as well).

**Comment by juanibiapina** [comment]
The same problem happens when using `fzf` for file completion. Both of these probably call `zle redisplay`, causing the issue. There is a workaround suggested here: https://superuser.com/questions/382503/how-can-i-put-a-newline-in-my-zsh-prompt-without-causing-terminal-redraw-issues

**Comment by segevfiner** [comment]
This feels like a zsh/zle bug. I'm not seeing this with zsh 5.7.1 but have seen it before. We should probably figure out which zsh version or commit fixed this.

**Comment by mcornella** [comment]
Latest versions of zsh don't have this bug, so I'll close this as an upstream issue. Confirmed to be fixed in 5.7.1 and buggy in 5.3, I don't know the oldest version with a fix.

**Comment by juanibiapina** [comment]
This is fixed for me on zsh 5.7.1, but not when inside tmux 3.0a.

**Comment by negebauer** [comment]
I had a similar issue and changing the locale fixed it https://github.com/ohmyzsh/ohmyzsh/issues/7558#issuecomment-460812339
